tests/tests_app/core/test_lightning_app_checkpoint.py

```diff
@@ -0,0 +1,146 @@
+import json
+import os
+from unittest import mock
+
+import pytest
+from tests_app import _PROJECT_ROOT
+
+from lightning_app import LightningApp, LightningFlow, LightningWork  # F401
+
+
+class SavedAppExample(LightningFlow):
+    def __init__(self):
+        super().__init__()
+        self.text = ""
+
+    def run(self):
+        print(f"what is the value of text? {self.text}")
+
+
+class DynamicWorkLevel3(LightningWork):
+    def __init__(self):
+        super().__init__()
+        self.text = ""
+
+    def run(self):
+        print(f"what is the value of text in the work? {self.text}")
+
+
+class DynamicFlowLevel2(LightningFlow):
+    def __init__(self):
+        super().__init__()
+        self.text = ""
+
+    def run(self):
+        if not getattr(self, "work_level_3", None):
+            self.work_level_3 = DynamicWorkLevel3()
+        self.work_level_3.run()
+
+    def load_state_dict(self, flow_state, children_states, strict) -> None:
+        self.work_level_3 = DynamicWorkLevel3()
```

Comment:
Could you try to see if overriding `state_dict` and add some extras required to re-instantiation works well ?